mongo_connector/doc_managers/neo4j_doc_manager.py

```diff
@@ -53,16 +58,73 @@ def upsert(self, doc, namespace, timestamp):
         "ns": namespace,
         "_ts": timestamp
     }
+    doc = self._formatter.format_document(doc)
+    self.store_nodes_and_relationships(doc, doc_type, doc_id)
+    
   
+  def store_nodes_and_relationships(self, doc, doc_type, doc_id):
+    self.build_nodes(doc_type, doc, doc_id)
+    for node_type in self.doc_types:
+      self.apply_id_constraint(node_type)
+    tx = self.graph.cypher.begin()
+    for statement in self.query_nodes.keys():
+      tx.append(statement, {"parameters":self.query_nodes[statement]})
+    main_type = self.doc_types.pop(0)
+    for node_type in self.doc_types:
+      tx.append(self.build_relationships(main_type, node_type), {"doc_id": doc_id})
+    tx.commit()
+    self.query_nodes = {}
+    self.doc_types = []
+
+  def build_nodes(self, doc_type, hash, id):
+    self.doc_types.append(doc_type)
+    parameters = {'_id':id}
+    for key in hash.keys():
+      if (type(hash[key]) is dict):
+        self.build_nodes(key, hash[key], id)
+      elif ((type(hash[key]) is list) and (type(hash[key][0]) is dict)):
+        for json in hash[key]:
+          json_key = key + str(hash[key].index(json))
+          self.build_nodes(json_key, json, id)
+      else:
+        value = hash[key]
+        parameters.update({ key: value })
+    query = "CREATE (c:Document:{doc_type} {{parameters}})".format(doc_type=doc_type)
+    self.query_nodes.update({query: parameters})
+
+  def build_relationships(self, main_type, node_type):
```

Comment:
I thought that refactor points should be added into a separate PR (i have started one for tests, general refactor and file split)